src/imago/src/file.rs

```
@@ -454,3 +456,58 @@ impl Display for File {
         }
     }
 }
+
+fn get_file_size(file: &fs::File) -> io::Result<u64> {
+    file.metadata().and_then(|m| {
+        #[cfg(windows)]
+        let is_block_device = false;
+        #[cfg(unix)]
+        let is_block_device = m.file_type().is_block_device();
```

Ok, so actually this is not a complete fix. It works on macOS with the buffered `/dev/disk*` but not `/dev/rdisk*`.
For one thing, I'd have to change the condition here to `is_block_device() || is_char_device()` but when I do that, I get `ERROR devices::virtio::block::worker] error processing request: WritingToDescriptor(Os { code: 22, kind: InvalidInput, message: "Invalid argument" })`.

I also noticed my disk sector size is 4096 but there's a lot of hardcoded 512 in the code. Raw disks don't allow unaligned access so that could be it...

Anyway, I'm not sure if I should attempt to fix this or rather ignore the character device case for now. Maybe return an error here.

---

I don’t think imago prescribes any block size. It queries the underlying block size and ensures requests are aligned to it (with read-modify-write cycles if necessary).

Imago’s users are responsible for choosing an appropriate block size, and imago reports the minimal underlying alignment via [FormatAccess::req_align()](https://docs.rs/imago/latest/imago/format/access/struct.FormatAccess.html#method.req_align) and [FormatAccess::mem_align()](https://docs.rs/imago/latest/imago/format/access/struct.FormatAccess.html#method.mem_align).

I assume the 512 comes from libkrun’s virtio-blk code, because 512 is the virtio-blk sector size. It’s possible to report other block sizes to the guest and thus force it to use those (via `virtio_blk_config.blk_size` and `virtio_blk_config.topology.physical_block_exp`).

---

Yeah, that's actually the code I was referring to. The virtio block worker in libkrun.

And I was wondering how it reports the virtualized block size into the guest. So thanks for pointers.

---

Ah, I think the problem is that libkrun opens the disk image without `direct_io` set, so the 0.1 version of imago assumes an alignment of 1. So backporting https://gitlab.com/hreitz/imago/-/commit/6aa125b06dc497dffec658b54f80d58f302535e5 to 0.1 could fix it.

---

@slp Another improvement might be to set `direct_io` to `true` in libkrun if the file path begins with `/dev/`. What do you think?

---

@nohajc I think in libkrun we should use `direct_io` unconditionally to avoid double-caching. @jakecorrenti @XanClic WDYT? Any reason to favor not using `direct_io`?

---

@nohajc Thanks a lot of opening the MR and for testing 6aa125b0. I can backport those to 0.1, no problem.

@slp I think `direct_io` is a good idea. As far as I understand, for QEMU, it is generally suggested to disable the host page cache on VM images to avoid double-caching, as you say. It may cause issues with images in locations that don’t support `O_DIRECT`, but the only case I know of was tmpfs, which now supports it, too.